utils_nlp/models/transformers/common.py

```diff
@@ -82,62 +99,125 @@ def get_default_optimizer(model, weight_decay, learning_rate, adam_epsilon):
         no_decay = ["bias", "LayerNorm.weight"]
         optimizer_grouped_parameters = [
             {
-                "params": [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)],
+                "params": [
+                    p
+                    for n, p in model.named_parameters()
+                    if not any(nd in n for nd in no_decay)
+                ],
                 "weight_decay": weight_decay,
             },
             {
-                "params": [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)],
+                "params": [
+                    p
+                    for n, p in model.named_parameters()
+                    if any(nd in n for nd in no_decay)
+                ],
                 "weight_decay": 0.0,
             },
         ]
-        optimizer = AdamW(optimizer_grouped_parameters, lr=learning_rate, eps=adam_epsilon)
+        optimizer = AdamW(
+            optimizer_grouped_parameters, lr=learning_rate, eps=adam_epsilon
+        )
         return optimizer
 
     @staticmethod
     def get_default_scheduler(optimizer, warmup_steps, num_training_steps):
         scheduler = get_linear_schedule_with_warmup(
-            optimizer, num_warmup_steps=warmup_steps, num_training_steps=num_training_steps
+            optimizer,
+            num_warmup_steps=warmup_steps,
+            num_training_steps=num_training_steps,
         )
         return scheduler
 
+    def prepare_model_and_optimizer(
+        self,
+        num_gpus,
+        gpu_ids,
+        local_rank,
+        weight_decay,
+        learning_rate,
+        adam_epsilon,
+        fp16=False,
+        fp16_opt_level="O1",
+        checkpoint_state_dict=None,
+    ):
+        """
+        This function initializes an optimizer and moves the model to a device.
+        It can be used by most child classes before calling fine_tune.
+        Child classes that require custom optimizers need to either override this
+            function or implement the steps listed below in the specified order
+            before fine-tuning.
+
+        The steps are performed in the following order:
+            1. Move model to device
+            2. Create optimizer
+            3. Initialize amp
+            4. Parallelize model
```

Comment:
Didn't realize the change has already merged into staging.  Can only comment here now. According to https://pytorch.org/tutorials/intermediate/ddp_tutorial.html, model needs to be parallelized before creating optimizer.

Reply:
never mind. Tried to parallelize model before initilize amp, got this error:   File "/dadendev/anaconda3/envs/nlp_gpu_transformer_bug_bash/lib/python3.6/site-packages/apex/amp/_initialize.py", line 75, in check_models
    "Parallel wrappers should only be applied to the model(s) AFTER \n"
RuntimeError: Incoming model is an instance of torch.nn.parallel.DataParallel. Parallel wrappers should only be applied to the model(s) AFTER
the model(s) have been returned from amp.initialize.